casper/src/test/scala/coop/rchain/casper/util/CasperUtilTest.scala

```diff
@@ -1,67 +1,60 @@
 package coop.rchain.casper.util
 
 import com.google.protobuf.ByteString
+import coop.rchain.casper.BlockGenerator
+import coop.rchain.casper.internals._
 import coop.rchain.casper.protocol._
-import coop.rchain.casper.protocol.Resource.ResourceClass.ProduceResource
 import org.scalatest.{FlatSpec, Matchers}
-import coop.rchain.crypto.hash.Sha256
+import scalaz.{State}
 
-import scala.collection.mutable
+import scala.collection.immutable.HashMap
 
-class CasperUtilTest extends FlatSpec with Matchers {
-  val resourceCounter: Iterator[Int] = Iterator.from(0)
+class CasperUtilTest extends FlatSpec with Matchers with BlockGenerator {
 
-  "isInMainChain" should "classify appropriately" in {
-    val chain = mutable.HashMap[ByteString, BlockMessage]()
-
-    val genesis: BlockMessage = createBlock(Seq())
-    chain += (genesis.blockHash -> genesis)
-
-    val blockMessage2: BlockMessage = createBlock(Seq(genesis.blockHash))
-    chain += (blockMessage2.blockHash -> blockMessage2)
+  type StateWithChain[A] = State[Chain, A]
 
-    val blockMessage3: BlockMessage = createBlock(Seq(blockMessage2.blockHash))
-    chain += (blockMessage3.blockHash -> blockMessage3)
-
-    isInMainChain(chain, genesis, blockMessage3) should be(true)
-    isInMainChain(chain, blockMessage2, blockMessage3) should be(true)
-    isInMainChain(chain, blockMessage3, blockMessage2) should be(false)
-    isInMainChain(chain, blockMessage3, genesis) should be(false)
+  "isInMainChain" should "classify appropriately" in {
+    val createChain =
+      for {
+        genesis <- createBlock[StateWithChain](Seq())
+        b2      <- createBlock[StateWithChain](Seq(genesis.blockHash))
+        b3      <- createBlock[StateWithChain](Seq(b2.blockHash))
+      } yield b3
+
+    val initState =
+      Chain(HashMap.empty[Int, BlockMessage], HashMap.empty[ByteString, BlockMessage], 0)
+    val (chain: Chain, lastBlock: BlockMessage) = createChain.run(initState)
+
+    val genesis = chain.idToBlocks(1)
+    val b2      = chain.idToBlocks(2)
+    val b3      = chain.idToBlocks(3)
+    isInMainChain(chain.hashToBlocks, genesis, b3) should be(true)
+    isInMainChain(chain.hashToBlocks, b2, b3) should be(true)
+    isInMainChain(chain.hashToBlocks, b3, b2) should be(false)
+    isInMainChain(chain.hashToBlocks, b3, genesis) should be(false)
```

Comment:
You should be able to move those `isInMainChain` assertions to your _for comprehension_ like this:
```
for {
…
_ = assert(isInMainChain(…))
…
} yield b3
```

Can you try and let me know?

Reply:
Well I need the entire chain to be built first

Reply:
Entire chain or only the parts that precede the blocks you want to test with `isInMainChain` ? Either way is fine, it's up to you.

Reply:
Yeah I'm thinking from the complete view of the "attacker" (basically the entire chain).